core/src/Kokkos_CopyViews.hpp

```diff
@@ -1077,17 +1080,37 @@ inline void deep_copy(
   }
 
   Kokkos::fence("Kokkos::deep_copy: scalar to scalar copy, pre copy fence");
-  if (dst.data() != src.data()) {
+
+  if constexpr (std::is_same_v<
+                    typename ViewTraits<DT, DP...>::value_type,
+                    typename ViewTraits<ST, SP...>::non_const_value_type>) {
+    if (dst.data() != src.data()) {
 #ifndef KOKKOS_ENABLE_IMPL_VIEW_LEGACY
-    using dst_ptr_type  = decltype(dst.data());
-    const size_t nbytes = allocation_size_from_mapping_and_accessor(
-                              src.mapping(), src.accessor()) *
-                          sizeof(std::remove_pointer_t<dst_ptr_type>);
+      using dst_ptr_type  = decltype(dst.data());
+      const size_t nbytes = allocation_size_from_mapping_and_accessor(
+                                src.mapping(), src.accessor()) *
+                            sizeof(std::remove_pointer_t<dst_ptr_type>);
 #else
-    const size_t nbytes = sizeof(value_type);
+      const size_t nbytes = sizeof(typename dst_type::value_type);
 #endif
-    Kokkos::Impl::DeepCopy<dst_memory_space, src_memory_space>(
-        dst.data(), src.data(), nbytes);
+      Kokkos::Impl::DeepCopy<dst_memory_space, src_memory_space>(
+          dst.data(), src.data(), nbytes);
+      Kokkos::fence(
+          "Kokkos::deep_copy: scalar to scalar copy, post copy fence");
+    }
```

Comment:
It's pre-existing that we only get a "Kokkos::deep_copy: scalar to scalar copy, pre copy fence" fence if the two views reference the same data. We could improve the fence message in that case, though.

Reply:
I don't want to explicitly say that it's a fence for copying between the same data. I think scalar to scalar is still fine.